_posts/tutorials/2021-05-20-frontends-with-javalinmithril.md

```diff
@@ -0,0 +1,450 @@
+         <type>jar</type>
+    </dependency>
+    <dependency>
+        <groupId>org.webjars.npm</groupId>
+        <artifactId>mithril</artifactId>
+        <version>2.0.4</version>
+    </dependency>
+    <dependency>
+        <groupId>com.fasterxml.jackson.core</groupId>
+        <artifactId>jackson-databind</artifactId>
+        <version>2.10.3</version>
+    </dependency>
+```
+
+<div class="comment" markdown="1">
+You can add all frontend dependencies as [Webjars](https://www.webjars.org/), which can be built directly from NPM.
+If something is available on NPM, it's also available as a Webjar, but as a prebuilt dist version (it has no dependencies).
+To view the full POM, please go to [GitHub](https://github.com/TareqK/javalinmithril-example/blob/master/pom.xml).
+</div>
+
+Now that we have all our dependencies in order, we need to configure our web server. Lets start off with 2 classes : the Main class, and the app class.
+
+Starting off, our Main class, in `/src/main/java/io/javalinmithril/demo/Main.java` is just a wrapper class(which I add because I am a sad non-kotlin boi :C  and I like neat class names too much C: ) that calls and starts our App
+
+```java
+import io.javalin.Javalin
+package io.javalin.javalinmithril.demo;
+
+public class Main {
+    
+    public static void main(String[] args){
+        new App().start();
+    }
+}
+```
+
+Our App class, which has the start method, looks like this
+
+```java
+...
+private Javalin app;
+
+public void start(){
+    app = Javalin.create();
+    app.config.enableWebjars();
+    JavalinMithril.configure(config -> {
+            config.isDev(true)
+                    .stateFunction((ctx) -> singletonMap("currentUser", currentUser(ctx)));
+    });
+    ...
+    app.start(7000);
+}
+...
+
+```
+Notice the `JavalinMithirl.configure` method. We are basically telling the plugin to work in dev mode, and giving it the state function that will be injected into each component - in this case, its the name of the logged in user
+
+We also need an HTML file to load our dependencies and to initialize Mithril. \\
+Let's create `/src/main/resources/mithril/layout.html`:
+
+```markup
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="utf8"/>
+        <script src="@cdnWebjar/mithril/2.0.4/mithril.min.js"></script>
+        <style>
+            ...
+        </style>
+        @componentRegistration
+    </head>
+    <body>
+
+    </body>
+    <script>
+        m.mount(document.body, {view:()=>@routeComponent})
+    </script>
+</html>
+```
+
+There are two Javalin specific things here: `@componentRegistration` and `@routeComponent`. The JavalinMithril plugin will scan your `/resources/mithril` folder and put all the dependencies you have imported in your current component into  into `@componentRegistration`, similar to how libraries are loaded via `<script>` tags. It will also will also let you choose one component to mount based on the current URL, this is the `@routeComponent`.
```

Comment:
```suggestion
There are two JavalinMithril specific things here: `@componentRegistration` and `@routeComponent`. The JavalinMithril plugin will scan your `/resources/mithril` folder and put all the dependencies you have imported in your current component into `@componentRegistration`, similar to how libraries are loaded via `<script>` tags. It will also will also let you choose one component to mount based on the current URL, this is the `@routeComponent`.
```